### DISCUSSION TITLE
Not getting user user given value for "Accept" header from swagger ui.

### DISCUSSION THREAD

**Discussion by prateek9623** [fastapi/fastapi#8556] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Header
from typing import Optional
import cv2
import io
import numpy as np
from fastapi.responses import StreamingResponse, JSONResponse

app = FastAPI()

@app.get("/image")
def get_image(accept: Optional[str] = Header("image/png")):
    accept_headers = accept.split(",")
    dummy = np.zeros((512, 512))
    if "image/png" in accept_headers or "*/*" in accept_headers:
        return StreamingResponse(
            io.BytesIO(cv2.imencode(".png", dummy)[1]), media_type="image/png"
        )
    elif "image/jpeg" in accept_headers:
        return StreamingResponse(
            io.BytesIO(cv2.imencode(".jpeg", dummy)[1]), media_type="image/jpeg"
        )
    else:
        return JSONResponse(
            status_code=406,
            content={
                "message": "encoding of type {} is not supported".format(accept)
            },
        )
```

### Description

Hi, I have an API that is sending responses in different encoding depending on the user and to detect which encoding server should use I am using the "Accept" header.  But when I tried to test it using Swagger UI, it always sent "application/json".

### Operating System

Linux, Windows

### FastAPI Version

0.70.1

### Python Version

Python 3.9.5

**Comment by STeveShary** [comment]
@prateek9623 What is the `curl` command that you get from using the SwaggerUI?

**Comment by prateek9623** [comment]
@STeveShary  ![image](https://user-images.githubusercontent.com/17100687/149832067-1874e547-7324-42c3-b2fe-c43267410101.png)

**Comment by STeveShary** [comment]
@prateek9623  Ok, the problem is that when you use Swagger, it is creating a HTTP `GET` request where you can only accept `application/json` content.  That is why you are getting your error.  The swagger UI is creating a request that will only accept JSON.

You are not declaring what type of response the request can get, so swagger is setting it as `application/json` by default.  I would suggest that you look at: https://fastapi.tiangolo.com/advanced/custom-response/ to set the correct response object.

This should modify the swagger UI to send the `accept` header of the correct encoding.

**Comment by prateek9623** [comment]
@STeveShary ok, but my issue is not with response, my issue is request header, it is not sent correctly. It should be sending image/png but it is sending application/json which i am assuming is default behaviour for "Accept" header. What i think is happening hear is, this default behaviour is overwriting user given "Accept" header.

**Comment by STeveShary** [comment]
You are correct that the request header is incorrect.  This can be corrected by defining the response type.

**Comment by prateek9623** [comment]
@STeveShary Hi, as per your suggestion I have assigned "StreamingResponse" as respone_class. But still "Accept" header is not user-given, now it is "*/*".
```python
from fastapi import FastAPI, Header
from typing import Optional
import cv2
import io
import numpy as np
from fastapi.responses import StreamingResponse, JSONResponse
from pydantic import BaseModel

app = FastAPI()


class Message(BaseModel):
    message: str

@app.get("/image", response_class= StreamingResponse, responses={"406": {"model":Message}})
def get_image(accept: Optional[str] = Header("image/png")):
    accept_headers = accept.split(",")
    dummy = np.zeros((512, 512))
    if "image/png" in accept_headers or "*/*" in accept_headers:
        return StreamingResponse(
            io.BytesIO(cv2.imencode(".png", dummy)[1]), media_type="image/png"
        )
    elif "image/jpeg" in accept_headers:
        return StreamingResponse(
            io.BytesIO(cv2.imencode(".jpeg", dummy)[1]), media_type="image/jpeg"
        )
    else:
        return JSONResponse(
            status_code=406,
            content={
                "message": "encoding of type {} is not supported".format(accept)
            },
        )
```
![image](https://user-images.githubusercontent.com/17100687/150193285-9325057a-d8ad-4a39-901a-8b61a48f610f.png)

**Comment by rafsaf** [comment]
**Observations:**
1. In this code, `accept` is overwritten by openapi docs as you observed:
    ```python
    @app.get("/image")
    def get_image(accept: str = Header("image/png")):
        return StreamingResponse(
            io.BytesIO("test".encode()), media_type=accept
        )
    ```
    BUT if in terminal you type
    ```curl
    curl -X 'GET' \
      'http://localhost:8000/image' \
      -H 'accept: image/png'
    ```
    It works just fine for me.

2. If then we change `accept` to `custom_accept`:
    ```python
    @app.get("/image")
    def get_image(custom_accept: str = Header("image/png")):
        return StreamingResponse(
            io.BytesIO("test".encode()), media_type=custom_accept
        )
    ```
    You get working `custom_accept` header just as expected and **docs are not broken**.

**Conclusions:**

It's not FastAPI related, not in any direct way I mean. I found something that may be interesting for you:
https://github.com/swagger-api/swagger-ui/issues/7374

That's kind of expected for me that your `accept` header will be overwritten by default one. 

So, with any luck, you have 3 options:
- either use custom header, in the end I think it's acceptable solution
- or read https://fastapi.tiangolo.com/advanced/extending-openapi/, read mentioned issue above and possible solution, extend/manipulate your OpenAPI schema and it should work.
- if you don't care about generated OpenAPI docs, just leave it as is, your endpoint will work, but docs will stay a bit not clean

**Comment by prateek9623** [comment]
i already knew this api work from source other than swagger, but it would have been nice if it also worked on swagger.
